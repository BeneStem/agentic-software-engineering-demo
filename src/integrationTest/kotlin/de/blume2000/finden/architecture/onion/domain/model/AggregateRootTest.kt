package de.blume2000.finden.architecture.onion.domain.model

import com.tngtech.archunit.base.DescribedPredicate.not
import com.tngtech.archunit.core.domain.JavaClass.Predicates.resideInAPackage
import com.tngtech.archunit.core.domain.JavaClass.Predicates.simpleNameEndingWith
import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.junit.AnalyzeClasses
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.fields
import de.blume2000.finden.architecture.onion.OnionTest
import de.blume2000.finden.architecture.onion.OnionTest.Companion.DOMAIN_MODEL_PACKAGE
import de.blume2000.finden.architecture.onion.domain.model.CommandTest.Companion.COMMAND
import de.blume2000.finden.architecture.onion.domain.model.EventTest.Companion.EVENT
import de.blume2000.finden.architecture.onion.domain.model.RepositoryTest.Companion.REPOSITORY
import org.jmolecules.ddd.annotation.AggregateRoot

@AnalyzeClasses(packages = [OnionTest.BOUNDED_CONTEXT_PACKAGE],
  importOptions = [ImportOption.DoNotIncludeTests::class, ImportOption.DoNotIncludeArchives::class])
class AggregateRootTest {

  @ArchTest
  fun aggregateRootClasses(classes: JavaClasses) =
    classes()
      .that().areAnnotatedWith(AggregateRoot::class.java)
      .should().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .check(classes)

  @ArchTest
  fun aggregateRootField(classes: JavaClasses) =
    fields()
      .that().areDeclaredInClassesThat().areAnnotatedWith(AggregateRoot::class.java)
      .should().beDeclaredInClassesThat().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .andShould().beFinal()
      .andShould().bePrivate()
      .andShould().haveRawType(resideInAPackage("$DOMAIN_MODEL_PACKAGE.").and(not(simpleNameEndingWith(REPOSITORY))
        .and(not(simpleNameEndingWith(COMMAND)))
        .and(not(simpleNameEndingWith(EVENT)))))
      .check(classes)
}
