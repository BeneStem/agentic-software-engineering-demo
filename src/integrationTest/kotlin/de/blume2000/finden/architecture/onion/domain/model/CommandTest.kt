package de.blume2000.finden.architecture.onion.domain.model

import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.junit.AnalyzeClasses
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.fields
import de.blume2000.finden.architecture.onion.OnionTest
import de.blume2000.finden.architecture.onion.OnionTest.Companion.DOMAIN_MODEL_PACKAGE

@AnalyzeClasses(packages = [OnionTest.BOUNDED_CONTEXT_PACKAGE],
  importOptions = [ImportOption.DoNotIncludeTests::class, ImportOption.DoNotIncludeArchives::class])
class CommandTest {

  companion object {
    const val COMMAND = "Command"
  }

  @ArchTest
  fun commandClasses(classes: JavaClasses) =
    classes()
      .that().haveSimpleNameEndingWith(COMMAND)
      .should().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .check(classes)

  @ArchTest
  fun commandFields(classes: JavaClasses) =
    fields()
      .that().areDeclaredInClassesThat().haveSimpleNameEndingWith(COMMAND)
      .and().doNotHaveName(OnionTest.COMPANION)
      .and().doNotHaveName(OnionTest.INSTANCE)
      .should().beDeclaredInClassesThat().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .andShould().beFinal()
      .andShould().bePrivate()
      .check(classes)
}
