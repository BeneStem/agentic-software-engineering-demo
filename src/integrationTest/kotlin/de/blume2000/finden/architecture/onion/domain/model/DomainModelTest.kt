package de.blume2000.finden.architecture.onion.domain.model

import com.tngtech.archunit.core.domain.JavaClass.Predicates.resideInAPackage
import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.domain.properties.HasType.Predicates.rawType
import com.tngtech.archunit.core.importer.ImportOption.DoNotIncludeArchives
import com.tngtech.archunit.core.importer.ImportOption.DoNotIncludeTests
import com.tngtech.archunit.junit.AnalyzeClasses
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import de.blume2000.finden.architecture.onion.OnionTest.Companion.BOUNDED_CONTEXT_PACKAGE
import de.blume2000.finden.architecture.onion.OnionTest.Companion.DOMAIN_MODEL_PACKAGE
import de.blume2000.finden.architecture.onion.domain.model.CommandTest.Companion.COMMAND
import de.blume2000.finden.architecture.onion.domain.model.EventTest.Companion.EVENT

@AnalyzeClasses(packages = [BOUNDED_CONTEXT_PACKAGE],
  importOptions = [DoNotIncludeTests::class, DoNotIncludeArchives::class])
class DomainModelTest {

  @ArchTest
  fun dddClasses(classes: JavaClasses) =
    classes()
      .that().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .and().haveSimpleNameNotEndingWith(COMMAND)
      .and().haveSimpleNameNotEndingWith(EVENT)
      .should().beAnnotatedWith(rawType(resideInAPackage("org.jmolecules.ddd.annotation")))
      .check(classes)
}
