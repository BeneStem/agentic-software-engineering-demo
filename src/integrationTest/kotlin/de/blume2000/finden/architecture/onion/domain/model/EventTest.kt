package de.blume2000.finden.architecture.onion.domain.model

import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.junit.AnalyzeClasses
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.fields
import de.blume2000.finden.architecture.onion.OnionTest
import de.blume2000.finden.architecture.onion.OnionTest.Companion.COMPANION
import de.blume2000.finden.architecture.onion.OnionTest.Companion.DOMAIN_MODEL_PACKAGE
import de.blume2000.finden.architecture.onion.OnionTest.Companion.INSTANCE

@AnalyzeClasses(packages = [OnionTest.BOUNDED_CONTEXT_PACKAGE],
  importOptions = [ImportOption.DoNotIncludeTests::class, ImportOption.DoNotIncludeArchives::class])
class EventTest {

  companion object {
    const val EVENT = "Event"
  }

  @ArchTest
  fun eventClasses(classes: JavaClasses) =
    classes()
      .that().haveSimpleNameEndingWith(EVENT)
      .should().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .check(classes)

  @ArchTest
  fun eventFields(classes: JavaClasses) =
    fields()
      .that().areDeclaredInClassesThat().haveSimpleNameEndingWith(EVENT)
      .and().doNotHaveName(COMPANION)
      .and().doNotHaveName(INSTANCE)
      .should().beDeclaredInClassesThat().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .andShould().beFinal()
      .andShould().bePrivate()
      .check(classes)
}
