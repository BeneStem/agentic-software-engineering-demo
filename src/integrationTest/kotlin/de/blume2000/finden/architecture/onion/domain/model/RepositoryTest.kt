package de.blume2000.finden.architecture.onion.domain.model

import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.junit.AnalyzeClasses
import com.tngtech.archunit.junit.ArchTest
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import de.blume2000.finden.architecture.onion.OnionTest
import de.blume2000.finden.architecture.onion.OnionTest.Companion.DOMAIN_MODEL_PACKAGE
import de.blume2000.finden.architecture.onion.OnionTest.Companion.PASSIVE_ADAPTER_PACKAGE
import org.jmolecules.ddd.annotation.Repository

@AnalyzeClasses(packages = [OnionTest.BOUNDED_CONTEXT_PACKAGE],
  importOptions = [ImportOption.DoNotIncludeTests::class, ImportOption.DoNotIncludeArchives::class])
class RepositoryTest {

  companion object {
    const val REPOSITORY = "Repository"
  }

  @ArchTest
  fun repositoryClasses(classes: JavaClasses) =
    classes()
      .that().haveSimpleNameEndingWith(REPOSITORY)
      .and().resideOutsideOfPackage("$PASSIVE_ADAPTER_PACKAGE.")
      .should().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .andShould().beAnnotatedWith(Repository::class.java)
      .andShould().beInterfaces()
      .check(classes)

  @ArchTest
  fun classesAnnotatedWithRepository(classes: JavaClasses) =
    classes()
      .that().areAnnotatedWith(Repository::class.java)
      .should().resideInAPackage("$DOMAIN_MODEL_PACKAGE.")
      .andShould().haveSimpleNameEndingWith(REPOSITORY)
      .andShould().beInterfaces()
      .check(classes)
}
