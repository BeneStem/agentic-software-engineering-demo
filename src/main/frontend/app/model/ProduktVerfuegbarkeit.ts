export class ProduktVerfuegbarkeit {
  bestellschluss: Date;
  liefertag: Date;

  constructor(bestellschluss: Date, liefertag: Date) {
    this.bestellschluss = bestellschluss;
    this.liefertag = liefertag;
  }
}

export type ProduktVerfuegbarkeitDTO = {
  bestellschluss: string;
  liefertag: string;
};
