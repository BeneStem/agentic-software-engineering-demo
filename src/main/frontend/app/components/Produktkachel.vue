<template>
  <a
    :href="produkt.produktUrl"
    class="erk_finden__produktkachel"
    @click="sendProduktClickTrackingEvent"
    :title="produkt.name"
  >
    <div class="erk_finden__produktkachel-bild" ref="htmlElement">
      <Produktbild :url="produkt.bildUrl" :altText="produkt.name" />
    </div>
    <div class="erk_finden__produktkachel-infos">
      <div class="erk_finden__produktkachel-name">{{ produkt.name }}</div>
      <div class="erk_finden__produktkachel-lieferaussage">
        <Fliesstext typ="klein">{{ lieferaussage }}</Fliesstext>
      </div>
      <div class="erk_finden__produktkachel-preis">
        {{ produkt.preis }} €
        <div class="erk_finden__produktkachel-streichpreis" v-if="produkt.zeigeStreichpreis">
          {{ produkt.streichpreis }} €
        </div>
      </div>
    </div>
  </a>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import Fliesstext from '@blume2000/design-system/dist/komponenten/content/Fliesstext.vue';
import { Produkt } from '../model/Produkt';
import Produktbild from './Produktbild.vue';
import { useProduktTracking } from '@/main/frontend/app/composables/ProduktTracking';
import { useProduktLieferaussage } from '@/main/frontend/app/composables/ProduktLieferaussage';

interface ProduktkachelProps {
  produkt: Produkt;
  position: number;
  listenname: string;
  listennummer: number;
}

const props = defineProps<ProduktkachelProps>();

const { lieferaussage } = useProduktLieferaussage(props.produkt.naechstmoeglicheVerfuegbarkeit);
const htmlElement = ref<HTMLElement | null>(null);
const { sendProduktClickTrackingEvent } = useProduktTracking(
  htmlElement,
  props.produkt.nummer,
  props.produkt.name,
  props.produkt.preis,
  props.listenname,
  props.listennummer,
  props.position,
  props.produkt.klassifikationName
);
</script>

<style lang="scss">
.erk_finden__produktkachel {
  display: block;
  color: $color-dark;
  text-decoration: none;
  outline: none;
  overflow: hidden;

  &-bild {
    border-radius: 6px;
    overflow: hidden;
    width: 100%;
    aspect-ratio: 1 / 1;

    img {
      width: 100%;
      outline: none;
    }
  }

  &-infos {
    margin-top: 16px;
    display: grid;
    grid-template-columns: 1fr max-content;
    grid-template-rows: 1fr max-content;
    grid-template-areas:
      'top-left top-right'
      'bottom-left bottom-right';
  }

  &-name {
    grid-area: top-left;
    word-break: break-word;
    @include font-BLUME2000(22, 20);
  }

  &-lieferaussage {
    grid-area: bottom-left;
  }

  &-preis {
    grid-column: top-right;
    grid-row-start: top-right;
    grid-row-end: bottom-right;
    justify-self: end;
    @include font-Rubik-medium;
    text-align: right;
    padding-left: 12px;
    padding-top: 2px;
  }

  &-streichpreis {
    margin-top: 4px;
    color: $color-alert;
    text-decoration: line-through;
    @include font-Rubik-light;
  }
}

@supports not (aspect-ratio: 1 / 1) {
  .erk_finden__produktkachel-bild::before {
    float: left;
    padding-top: 100%;
    content: '';
  }

  .erk_finden__produktkachel-bild::after {
    display: block;
    content: '';
    clear: both;
  }
}
</style>
