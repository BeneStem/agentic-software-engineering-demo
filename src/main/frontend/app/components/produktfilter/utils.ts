import { FilterTyp } from '@/main/frontend/app/model/FilterTyp';

export function erstelleFilterTabId(filterTyp: FilterTyp): `filter-tab-${FilterTyp}` {
  return `filter-tab-${filterTyp}`;
}

export function erstelleFilterTabPanelId(filterTyp: FilterTyp): `filter-tabpanel-${FilterTyp}` {
  return `filter-tabpanel-${filterTyp}`;
}
