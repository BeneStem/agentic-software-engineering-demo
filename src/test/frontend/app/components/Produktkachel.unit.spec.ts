import { mount, shallowMount } from '@vue/test-utils';
import Produktkachel from '@/main/frontend/app/components/Produktkachel.vue';
import { testProdukt, testProduktMitStreichpreis } from '@/test/frontend/utils/TestProduktHelper';
import { ProduktVerfuegbarkeit } from '@/main/frontend/app/model/ProduktVerfuegbarkeit';
import { addDays, addHours, format, subHours } from 'date-fns';
import deLocale from 'date-fns/locale/de';

describe('Produktkachel.vue', () => {
  require('vue-router').useRoute = jest.fn();
  require('vue-router').useRoute.mockImplementation(() => ({ path: '/test' }));

  const position = 1;
  const listenname = 'listenname';
  const listennummer = 1337;

  it('Rendert eine Produktkachel', () => {
    // Given
    const wrapper = shallowMount(Produktkachel, {
      props: {
        produkt: testProdukt,
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Rendert eine Produktkachel mit Streichpreis', () => {
    // Given
    const wrapper = shallowMount(Produktkachel, {
      props: {
        produkt: testProduktMitStreichpreis,
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Rendert eine Produktkachel mit Verfügbarkeit heute bestellt, morgen da', () => {
    // Given
    const naechstmoeglicheVerfuegbarkeit: ProduktVerfuegbarkeit = {
      bestellschluss: addHours(new Date(), 2),
      liefertag: addDays(new Date(), 1),
    };
    const wrapper = mount(Produktkachel, {
      props: {
        produkt: {
          ...testProdukt,
          naechstmoeglicheVerfuegbarkeit,
        },
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.find('.erk_finden__produktkachel-lieferaussage *').text()).toEqual(
      'Heute bestellt - morgen geliefert'
    );
  });

  it('Rendert eine Produktkachel mit Verfügbarkeit heute bestellt, übermorgen da', () => {
    // Given
    const naechstmoeglicheVerfuegbarkeit: ProduktVerfuegbarkeit = {
      bestellschluss: addHours(new Date(), 2),
      liefertag: addDays(new Date(), 2),
    };
    const wrapper = mount(Produktkachel, {
      props: {
        produkt: {
          ...testProdukt,
          naechstmoeglicheVerfuegbarkeit,
        },
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.find('.erk_finden__produktkachel-lieferaussage *').text()).toEqual(
      'Heute bestellt - übermorgen geliefert'
    );
  });

  it('Rendert eine Produktkachel mit Verfügbarkeit heute bestellt, ab XX.XX da', () => {
    // Given
    const naechstmoeglicheVerfuegbarkeit: ProduktVerfuegbarkeit = {
      bestellschluss: addHours(new Date(), 2),
      liefertag: addDays(new Date(), 4),
    };
    const wrapper = mount(Produktkachel, {
      props: {
        produkt: {
          ...testProdukt,
          naechstmoeglicheVerfuegbarkeit,
        },
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.find('.erk_finden__produktkachel-lieferaussage *').text()).toEqual(
      `Heute bestellt - ab ${format(naechstmoeglicheVerfuegbarkeit.liefertag, 'eeeeee, dd.MM.', {
        locale: deLocale,
      })} geliefert`
    );
  });

  it('Rendert eine Produktkachel ohne Verfügbarkeit falls Verfügbarkeit nicht mehr gegeben', () => {
    // Given
    const naechstmoeglicheVerfuegbarkeit: ProduktVerfuegbarkeit = {
      bestellschluss: subHours(new Date(), 2),
      liefertag: addDays(new Date(), 2),
    };
    const wrapper = mount(Produktkachel, {
      props: {
        produkt: {
          ...testProdukt,
          naechstmoeglicheVerfuegbarkeit,
        },
        position,
        listenname,
        listennummer,
      },
    });

    // Then
    expect(wrapper.find('.erk_finden__produktkachel-lieferaussage *').text()).toEqual('');
  });

  it('Sendet beim Klick auf die Kachel ein Event an GA', () => {
    // Given
    const wrapper = shallowMount(Produktkachel, {
      props: {
        produkt: testProduktMitStreichpreis,
        position,
        listenname,
        listennummer,
      },
    });

    wrapper.vm.sendProduktClickTrackingEvent = jest.fn();

    // When
    wrapper.find('a').trigger('click');

    // Then
    expect(wrapper.vm.sendProduktClickTrackingEvent).toHaveBeenCalled();
  });
});
