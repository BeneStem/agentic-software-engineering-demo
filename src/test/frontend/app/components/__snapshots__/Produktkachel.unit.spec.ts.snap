// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Produktkachel.vue Rendert eine Produktkachel 1`] = `
<a
  class="erk_finden__produktkachel"
  href="#"
  title="Produktname"
>
  <div
    class="erk_finden__produktkachel-bild"
  >
    <produktbild-stub
      alttext="Produktname"
      url="https://assets.ecom.blume2000.de/images/produkt_platzhalter.png"
    />
  </div>
  <div
    class="erk_finden__produktkachel-infos"
  >
    <div
      class="erk_finden__produktkachel-name"
    >
      Produktname
    </div>
    <div
      class="erk_finden__produktkachel-lieferaussage"
    >
      <fliesstext-stub
        theme="hell"
        typ="klein"
      />
    </div>
    <div
      class="erk_finden__produktkachel-preis"
    >
      10.00 € 
      <!--v-if-->
    </div>
  </div>
</a>
`;

exports[`Produktkachel.vue Rendert eine Produktkachel mit Streichpreis 1`] = `
<a
  class="erk_finden__produktkachel"
  href="#"
  title="Produktname"
>
  <div
    class="erk_finden__produktkachel-bild"
  >
    <produktbild-stub
      alttext="Produktname"
      url="https://assets.ecom.blume2000.de/images/produkt_platzhalter.png"
    />
  </div>
  <div
    class="erk_finden__produktkachel-infos"
  >
    <div
      class="erk_finden__produktkachel-name"
    >
      Produktname
    </div>
    <div
      class="erk_finden__produktkachel-lieferaussage"
    >
      <fliesstext-stub
        theme="hell"
        typ="klein"
      />
    </div>
    <div
      class="erk_finden__produktkachel-preis"
    >
      10.00 € 
      <div
        class="erk_finden__produktkachel-streichpreis"
      >
        20.00 € 
      </div>
    </div>
  </div>
</a>
`;
